---
title: "02-25_932Hw1"
author: "Erika Sanchez"
date: "2/25/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Version-controlled directory system
   https://github.com/esanchezb/esb-agro932.git
 
# Download Phaseolus reference genome
#### from https://plants.ensembl.org/Phaseolus_vulgaris/Info/Index
```{bash, eval=FALSE}
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/phaseolus_vulgaris/dna/Phaseolus_vulgaris.PhaVulg1_0.dna_sm.nonchromosomal.fa.gz
```

#### Unzip reference genome file
```{bash, eval=FALSE}
gunzip Phaseolus_vulgaris.PhaVulg1_0.dna_sm.nonchromosomal.fa.gz 

mv Phaseolus_vulgaris.PhaVulg1_0.dna_sm.nonchromosomal.fa PhaVulg1_0.fa

bwa index PhaVulg1_0.fa
```

# 2. Simulation strategy and the hypothesis to test (positive, negative, or neutral selection)

- Reference (about 6.268 kb)
    PhaVulg1_0.fa
- 20x coverage
    N 5000
- PE 100bp
    -1 100 -2 100
- Only SNP no Indel
    -R 0 -X 0
-	Mutation rate is low
    -r 0.01

#### simulation
```{bash, eval=FALSE}
for i in {1..10}; do    wgsim ../../largedata/PhaVulg1_0.fa -e 0 -d 500 -N 5000 -1 100 -2 100 -r 0.01  -R 0 -X 0 l$i.read1.fq l$i.read2.fq; done
```

#### count reads
```{bash, eval=FALSE}
wc -l l7.read2.fq
```

# Align the NGS reads to the reference genome

```{bash, eval=FALSE}
module load bwa samtools

# index the reference genome
bwa index Phaseolus_vulgaris.v1_0.fa

# bwa mem to align the reads to the reference genome 
for i in {1..10}; do bwa mem ../../largedata/PhaVulg1_0.fa l$i.read1.fq l$i.read2.fq | samtools view -bSh - > l$i.bam; done

# sort 
for i in *.bam; do samtools sort $i -o sorted_$i; done

# index ("Indexing a genome sorted BAM file allows one to quickly extract alignments overlapping particular genomic regions")
for i in sorted*.bam; do samtools index $i; done

# samtools flagstat â€“ counts the number of alignments for each FLAG type
samtools flagstat sorted_l1.bam
```

```{bash, eval=FALSE}
# write the bam files to a txt file
mkdir bam_files
mv sorted*.bam bam_files
cd bam_files
ls sorted*.bam > bam.txt
```

```{bash, eval=FALSE}
# run angsd
../../../bin/angsd/angsd -bam bam.txt -doSaf 1 -anc ../../../largedata/PhaVulg1_0.fa -GL 1  -out out_ang

../../../bin/angsd/misc/realSFS out_ang.saf.idx > out.sfs

cp out.sfs ../../cache/
```

```{bash, eval=FALSE}
#first generate .saf file [alternative to see more data - but I obtained the same]
../../../bin/angsd/angsd -bam bam.txt -doSaf 1 -out smallFolded -anc ../../../largedata/PhaVulg1_0.fa -GL 2
#now try the EM optimization with 4 threads
../../../bin/angsd/misc/realSFS smallFolded.saf.idx -maxIter 100 -P 4 > smallFolded.sfs
#in R
sfs<-scan("cache/smallFolded.sfs")
barplot(sfs[-1])
```


# Calculate $\theta$ values
```{bash, eval=FALSE}
../../../bin/angsd/angsd -bam bam.txt -out out_ang -doThetas 1 -doSaf 1 -pest out.sfs -anc ../../../largedata/PhaVulg1_0.fa -GL 1

../../../bin/angsd/misc/thetaStat print out_ang.thetas.idx > theta.txt

cp theta.txt ../../cache/
```

# Calculate $F_{ST}$ values
```{bash, eval=FALSE}
# pop1 = l1-l5
# pop2 = l5-l10
cp bam.txt pop1.txt
cp bam.txt pop2.txt

../../../bin/angsd/angsd -b pop1.txt -anc ../../../largedata/PhaVulg1_0.fa -out pop1 -dosaf 1 -gl 1

../../../bin/angsd/angsd -b pop2.txt -anc ../../../largedata/PhaVulg1_0.fa -out pop2 -dosaf 1 -gl 1

# calculate the 2dsfs prior
../../../bin/angsd/misc/realSFS pop1.saf.idx pop2.saf.idx > pop1.pop2.ml

# prepare the fst for easy window analysis
../../../bin/angsd/misc/realSFS fst index pop1.saf.idx pop2.saf.idx -sfs pop1.pop2.ml -fstout out

# get the global estimate
../../../bin/angsd/misc/realSFS fst stats out.fst.idx

# Result: FST.Unweight[nObs:1771942]:0.049392 Fst.Weight:0.049273

# below is not tested that much, but seems to work
../../../bin/angsd/misc/realSFS fst stats2 out.fst.idx -win 500 -step 100 > fst_win.txt

## copy sfs to cache
cp fst_win.txt ../../cache/
```

```{bash, eval=FALSE}
# bigger window - small output file 
../../../bin/angsd/misc/realSFS fst stats2 out.fst.idx -win 5000 -step 1000 > fst2_win.txt
## copy sfs to cache
cp fst2_win.txt ../../cache/
```

# module load R
```{bash, eval=FALSE}
# module load R
# R
s <- scan('out.sfs') 

# R studio
s <- scan('cache/out.sfs')  #read in the log sfs
s <- s[-c(1,length(s))] #plot variable sites
s <- s/sum(s)
barplot(s,names=1:length(s), main='SFS')
```

## Histogram distribution of the theta values
```{r, eval=FALSE}
theta <- read.table("cache/theta.txt", header=FALSE)
colnames(theta)
# last sentence in R window: [ reached 'max' / getOption("max.print") -- omitted 3517353 rows ]
theta2 <- read.table("theta.txt", header=FALSE)
hist(theta$V4, main=Theta-Pairwise)
```

## Scatter plot of the Fst values
```{r, eval=FALSE}
fst <- read.table("cache/fst_win.txt", skip=1, header=FALSE)
names(fst)[c(3,5)] <- c("midp", "fst")
plot(fst$midp, fst$fst, xlab="Physical position", ylab="Fst", col="#a0615f", pch=16)
```

# 4. General feature format (GFF) from EnsemblPlant

```{bash, eval=FALSE}
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/gff3/phaseolus_vulgaris/Phaseolus_vulgaris.PhaVulg1_0.46.gff3.gz
gunzip Phaseolus_vulgaris.PhaVulg1_0.46.gff3.gz
```

# R to process the GFF3 file
```{r, eval=FALSE}
# install.package("data.table")
library("data.table")

## simply read in wouldn't work
gff <- fread("largedata/Phaseolus_vulgaris.PhaVulg1_0.46.gff3", skip="#", header=FALSE, data.table=FALSE) #?
gff <- data.table("largedata/Phaseolus_vulgaris.PhaVulg1_0.46.gff3", skip="##", header=FALSE) #putty

## grep -v means select lines that not matching any of the specified patterns
gff <- fread(cmd='grep -v "#" largedata/Phaseolus_vulgaris.PhaVulg1_0.46.gff3', header=FALSE, data.table=FALSE)

names(gff) <- c("seq", "source", "feature", "start", "end", "score", "strand", "phase", "att")
table(gff$feature)
```

## Get genes and upstream and downstream 5kb regions

```{r, eval=FALSE}
g <- subset(gff, feature %in% "gene")
g$geneid <- gsub(".*gene:|;biotype.*", "", g$att)
### + strand
gp <- subset(g, strand %in% "+") 
# nrow(gp) 14141

### get the 5k upstream of the + strand gene model
gp_up <- gp
gp_up$end <- gp_up$start - 1
gp_up$start <- gp_up$end - 5000 

### get the 5k downstream of the + strand gene model
gp_down <- gp
gp_down$start <- gp_down$end + 1
gp_down$end <- gp_down$start + 5000 
```

## Get genes upstream and downstream 5kb regions

```{r echo=FALSE, eval=FALSE}
g <- subset(gff, feature %in% "gene")
g$geneid <- gsub(".*gene:|;biotype.*", "", g$att)
### - strand
gm <- subset(g, strand %in% "-") 
# nrow(gm) 13993

### get the 5k upstream of the + strand gene model
gm_up <- gm
gm_up$start <- gm_up$end + 1
gm_up$end <- gm_up$start + 5000 

### get the 5k downstream of the + strand gene model
gm_down <- gm
gm_down$end <- gm_down$start - 1
gm_down$start <- gm_down$end - 5000 

gup <- rbind(gp_up, gm_up)

fwrite(gup, "esb-agro932/cache/mt_gene_up5k.txt", sep="\t", row.names = FALSE, quote=FALSE)

gdown <- rbind(gp_down, gm_down)

fwrite(gdown, "esb-agro932/cache/mt_gene_down5k.txt", sep="\t", row.names = FALSE, quote=FALSE)
```

```{r, eval=FALSE}
### - strand
gm <- subset(g, strand %in% "-") 
dim(gm) # 13993    10
fwrite(g, "esb-agro932/cache/mt_gene.txt", sep="\t", row.names = FALSE, quote=FALSE)
```

Interpret theta results
```{r, eval=FALSE}
library("data.table")
library("GenomicRanges")
library("plyr")

theta <- fread("esb-agro932/cache/theta.txt", data.table=FALSE) # Putty
theta <- data.table("cache/theta.txt", data.table=FALSE) # Local
names(theta)[1] <- "seq"
up5k <- data.table("cache/mt_gene_up5k.txt", header=TRUE) # local
up5k <- data.table("esb-agro932/cache/mt_gene_up5k.txt", header=TRUE) # Putty doesn't work
up5k <- read.table("esb-agro932/cache/mt_gene_up5k.txt", header=TRUE) # Putty doesn't work

up5k <- read.delim("esb-agro932/cache/mt_gene_up5k.txt", header=TRUE) # read 10 variables


### define the subject file for theta values
grc <- with(theta, GRanges(seqnames=seq, IRanges(start=Pos, end=Pos)))


######## CONTINUE HERE
### define the query file for genomic feature
grf <- with(up5k, GRanges(seqnames=seq, IRanges(start=start, end=end), geneid=geneid))
    
### find overlaps between the two
tb <- findOverlaps(query=grf, subject=grc)
tb <- as.matrix(tb)
    
out1 <- as.data.frame(grf[tb[,1]])
out2 <- as.data.frame(grc[tb[,2]])

### for each genomic feature, find the sites with non-missing data
out <- cbind(out1, out2[, "start"]) 
names(out)[ncol(out)] <- "pos"
```

```{r, eval=FALSE}
#define unique identifier and merge with the thetas
out$uid <- paste(out$seqnames, out$pos, sep="_")
theta$uid <- paste(theta$seq, theta$Pos, sep="_")
df <- merge(out, theta[, c(-1, -2)], by="uid")

# for each upstream 5k region, how many theta values
###ddply: This function splits data frames by variables and combines the result into a data frame
mx <- ddply(df, .(geneid), summarise,  
            Pairwise = mean(Pairwise, na.rm=TRUE),
            thetaH = mean(thetaH, na.rm=TRUE),
            nsites = length(uid))
```


Copy and paste everything above, and pack it into an `R` function:
```{r, eval=FALSE}
get_mean_theta <- function(gf_file="esb-agro932/cache/mt_gene_up5k.txt"){
  
  # gf_file: gene feature file [chr, ="cache/mt_gene_up5k.txt"]
  
  theta <- fread("esb-agro932/cache/theta.txt", data.table=FALSE)
  names(theta)[1] <- "seq"
  up5k <- read.delim(gf_file, header=TRUE)

  ### define the subject file for theta values
  grc <- with(theta, GRanges(seqnames=seq, IRanges(start=Pos, end=Pos)))
  ### define the query file for genomic feature
  grf <- with(up5k, GRanges(seqnames=seq, IRanges(start=start, end=end), geneid=geneid))
    
  ### find overlaps between the two
  tb <- findOverlaps(query=grf, subject=grc)
  tb <- as.matrix(tb)
    
  out1 <- as.data.frame(grf[tb[,1]])
  out2 <- as.data.frame(grc[tb[,2]])
  ### for each genomic feature, find the sites with non-missing data
  out <- cbind(out1, out2[, "start"]) 
  names(out)[ncol(out)] <- "pos"
  #define unique identifier and merge with the thetas
  out$uid <- paste(out$seqnames, out$pos, sep="_")
  theta$uid <- paste(theta$seq, theta$Pos, sep="_")
  df <- merge(out, theta[, c(-1, -2)], by="uid")
  # for each upstream 5k region, how many theta values
  mx <- ddply(df, .(geneid), summarise,
            Pairwise = mean(Pairwise, na.rm=TRUE),
            thetaH = mean(thetaH, na.rm=TRUE),
            nsites = length(uid))
  return(mx)
}
```

since function didn't work I'll run separate the down5k
```{r, eval=FALSE}

  theta <- fread("esb-agro932/cache/theta.txt", data.table=FALSE)
  names(theta)[1] <- "seq"
  up5k <- read.delim("esb-agro932/cache/mt_gene_down5k.txt", header=TRUE)
  
  ### define the subject file for theta values
  grcdw <- with(theta, GRanges(seqnames=seq, IRanges(start=Pos, end=Pos)))
  ### define the query file for genomic feature
  grfdw <- with(up5k, GRanges(seqnames=seq, IRanges(start=start, end=end), geneid=geneid))
    
  ### find overlaps between the two
  tbdw <- findOverlaps(query=grf, subject=grc)
  tbdw <- as.matrix(tb)
    
  out1dw <- as.data.frame(grf[tb[,1]])
  out2dw <- as.data.frame(grc[tb[,2]])
  ### for each genomic feature, find the sites with non-missing data
  outdw <- cbind(out1dw, out2dw[, "start"]) 
  names(outdw)[ncol(outdw)] <- "pos"
  #define unique identifier and merge with the thetas
  outdw$uid <- paste(outdw$seqnames, outdw$pos, sep="_")
  theta$uid <- paste(theta$seq, theta$Pos, sep="_")
  df <- merge(outdw, theta[, c(-1, -2)], by="uid")
  # for each upstream 5k region, how many theta values
  mxdown <- ddply(df, .(geneid), summarise,
            Pairwise = mean(Pairwise, na.rm=TRUE),
            thetaH = mean(thetaH, na.rm=TRUE),
            nsites = length(uid))
  

```
## Plot the results

Run the customized R function

```{r, eval=FALSE}
### apply the function
up5k <- get_mean_theta(gf_file="esb-agro932/cache/mt_gene_up5k.txt")
down5k <- get_mean_theta(gf_file="esb-agro932/cache/mt_gene_down5k.txt")
```

And then plot the results:
```{r, eval=FALSE}
library("ggplot2")
up5k$feature <- "up 5k"
down5k$feature <- "down 5k"
res <- rbind(up5k, down5k)
ggplot(res, aes(x=feature, y=Pairwise, fill=feature)) + 
  geom_violin(trim=FALSE)+
  labs(title="Theta value", x="", y = "Log10 (theta)")+
  geom_boxplot(width=0.1, fill="white")+
  scale_fill_brewer(palette="Blues") + 
  theme_classic()

```

Second plot (Santos)
```{r, eval=FALSE}
library("ggplot2")
up5k_f <- get_mean_Fst(gf_file="esb-agro932/cache/mt_gene_up5k.txt")
down5k_f <- get_mean_Fst(gf_file="esb-agro932/cache/mt_gene_down5k.txt")
gene_f <- get_mean_Fst(gf_file="esb-agro932/cache/mt_gene.txt")

up5k_f$feature <- "up 5k"
down5k_f$feature <- "down 5k"
gene_f$feature <- "genic"
res_f <- rbind(up5k_f, down5k_f)
res_nongenic_f <- res_f
res_nongenic_f$feature <- "intergenic"
res_t_f <- rbind(res_f, gene_f, res_nongenic_f)
ggplot(res_t_f, aes(x=feature, y=Fst, fill=feature)) + 
  geom_violin(trim=FALSE)+
  labs(title="Fst distribution", x="", y = "Fst")+
  geom_boxplot(width=0.1, fill="white")+
  scale_fill_brewer(palette="greens") + 
  theme_classic()
```
